src/google/protobuf/compiler/csharp/csharp_options.h

```diff
@@ -41,14 +41,29 @@ namespace csharp {
 
 // Generator options (used by csharp_generator.cc):
 struct Options {
-  Options() : file_extension(".cs"), base_namespace(""), generate_directories(false) {
+  Options() :
+      file_extension(".cs"),
+      base_namespace(""),
+      base_namespace_specified(false) {
   }
   // Extension of the generated file. Defaults to ".cs"
   string file_extension;
-  // Base namespace to use to create directory hierarchy. Defaults to ""
+  // Base namespace to use to create directory hierarchy. Defaults to "".
+  // This option allows the simple creation of a conventional C# file layout,
+  // where directories are created relative to a project-specific base
+  // namespace. For example, in a project with a base namespace of PetShop, a
+  // proto of user.proto with a C# namespace of PetShop.Model.Shared would
+  // generate Model/Shared/User.cs underneath the specified --csharp_out
+  // directory.
+  //
+  // If no base namespace is specified, all files are generated in the
+  // --csharp_out directory, with no subdirectories created automatically.
```

Comment:
> Does this mean that two files foo/user.proto and bar/user.proto would conflict and overwrite the same output file if base_namespace_specified=false? This seems bad.

That's true, but is also the behaviour for js. Given that this is something the consumer (rather than just the proto producer) can control, I'm not too worried, to be honest.

> Is there a compelling use case for the base_namespace_specified=false behavior?

For users who just want a single place for their generated code, and don't particularly want to follow the layout convention (which is only a convention, not enforced), I think it can make sense. It also means that in simple cases - particularly when learning to start with - you don't need the extra command-line option.

Even for production use, I can see some reasonably small projects having messages which are all in the same namespace, picking the right output directory, and everything just works without the extra option.